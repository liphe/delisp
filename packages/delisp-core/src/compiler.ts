import runtime from "@delisp/runtime";
import {
  Expression,
  isDefinition,
  isExport,
  isTypeAlias,
  Module,
  SConditional,
  SDefinition,
  SExport,
  SFunction,
  SFunctionCall,
  SIdentifier,
  SLet,
  SRecord,
  SVectorConstructor,
  Syntax
} from "./syntax";

import { methodCall } from "./compiler/estree-utils";
import { last, mapObject, maybeMap } from "./utils";

import { InvariantViolation } from "./invariant";

import { printHighlightedExpr } from "./error-report";

import {
  DefinitionBackend,
  dynamicDefinition,
  staticDefinition
} from "./compiler/definitions";
import { cjs, esm, ModuleBackend } from "./compiler/modules";

import { member } from "./compiler/estree-utils";
import {
  compileInlinePrimitive,
  isInlinePrimitive
} from "./compiler/inline-primitives";
import { identifierToJS, isValidJSIdentifierName } from "./compiler/jsvariable";
import { pprint } from "./printer";

import * as escodegen from "escodegen";
import * as JS from "estree";

import createDebug from "debug";
const debug = createDebug("delisp:compiler");

interface EnvironmentBinding {
  jsname: string;
  source: "lexical" | "module" | "primitive";
}

export interface Environment {
  defs: DefinitionBackend;
  moduleFormat: ModuleBackend;
  bindings: {
    [symbol: string]: EnvironmentBinding;
  };
}

export interface CompilerOptions {
  definitionContainer?: string;
  esModule?: boolean;
}

function addBinding(varName: string, env: Environment): Environment {
  return {
    ...env,
    bindings: {
      ...env.bindings,
      [varName]: { jsname: identifierToJS(varName), source: "lexical" }
    }
  };
}

function lookupBinding(varName: string, env: Environment) {
  return env.bindings[varName];
}

function compileBody(body: Expression[], env: Environment): JS.BlockStatement {
  const middleForms = body.slice(0, -1);
  const lastForm = last(body)!;

  return {
    type: "BlockStatement",
    body: [
      ...middleForms.map(
        (e): JS.ExpressionStatement => ({
          type: "ExpressionStatement",
          expression: compile(e, env)
        })
      ),
      {
        type: "ReturnStatement",
        argument: compile(lastForm, env)
      }
    ]
  };
}

// Convert a Delisp variable name to Javascript. This function should
// be injective so there is no collisions and the output should be a
// valid variable name.

function compileLambda(
  fn: SFunction,
  env: Environment
): JS.ArrowFunctionExpression {
  const newEnv = fn.lambdaList.positionalArgs.reduce(
    (e, param) => addBinding(param.name, e),
    env
  );

  const jsargs = fn.lambdaList.positionalArgs.map(
    (param): JS.Pattern => ({
      type: "Identifier",
      name: lookupBinding(param.name, newEnv).jsname
    })
  );

  const implicitReturn = fn.body.length === 1;

  const body: JS.Expression | JS.Statement = implicitReturn
    ? compile(fn.body[0], newEnv)
    : compileBody(fn.body, newEnv);

  return {
    type: "ArrowFunctionExpression",
    params: [...jsargs],
    body,
    expression: implicitReturn
  };
}

function compileDefinition(def: SDefinition, env: Environment): JS.Statement {
  const value = compile(def.value, env);
  const name = lookupBinding(def.variable, env).jsname;
  return env.defs.define(name, value);
}

function compileFunctionCall(
  funcall: SFunctionCall,
  env: Environment
): JS.Expression {
  const compiledArgs = funcall.args.map(arg => compile(arg, env));
  if (funcall.fn.tag === "identifier" && isInlinePrimitive(funcall.fn.name)) {
    return compileInlinePrimitive(funcall.fn.name, compiledArgs, "funcall");
  } else {
    return {
      type: "CallExpression",
      callee: compile(funcall.fn, env),
      arguments: compiledArgs
    };
  }
}

function compileIdentifier(ref: SIdentifier, env: Environment): JS.Expression {
  const binding = lookupBinding(ref.name, env);

  if (!binding) {
    if (isInlinePrimitive(ref.name)) {
      return compileInlinePrimitive(ref.name, [], "value");
    } else {
      return env.defs.access(identifierToJS(ref.name));
    }
  }

  switch (binding.source) {
    case "primitive":
      return member(
        {
          type: "Identifier",
          name: "env"
        },
        binding.jsname
      );
    case "module":
      return env.defs.access(binding.jsname);
    case "lexical":
      return {
        type: "Identifier",
        name: binding.jsname
      };
    default:
      throw new InvariantViolation(
        "This switch-statement should be exhaustive but TS doesn't detect it somehow."
      );
  }
}

function compileConditional(
  expr: SConditional,
  env: Environment
): JS.Expression {
  return {
    type: "ConditionalExpression",
    test: compile(expr.condition, env),
    consequent: compile(expr.consequent, env),
    alternate: compile(expr.alternative, env)
  };
}

function compileLetBindings(expr: SLet, env: Environment): JS.Expression {
  const newenv = expr.bindings.reduce(
    (e, binding) => addBinding(binding.var, e),
    env
  );

  return {
    type: "CallExpression",
    callee: {
      type: "FunctionExpression",
      params: expr.bindings.map(
        (b): JS.Pattern => ({
          type: "Identifier",
          name: lookupBinding(b.var, newenv).jsname
        })
      ),
      body: compileBody(expr.body, newenv)
    },
    arguments: expr.bindings.map(b => compile(b.value, env))
  };
}

function literal(value: number | string): JS.Literal {
  return {
    type: "Literal",
    value
  };
}

function identifier(name: string): JS.Identifier {
  return {
    type: "Identifier",
    name
  };
}

function compileVector(
  expr: SVectorConstructor,
  env: Environment
): JS.Expression {
  return {
    type: "ArrayExpression",
    elements: expr.values.map(e => compile(e, env))
  };
}

function compileRecord(expr: SRecord, env: Environment): JS.Expression {
  const newObj: JS.ObjectExpression = {
    type: "ObjectExpression",
    properties: expr.fields.map(
      ({ label, value }): JS.Property => {
        if (!label.startsWith(":")) {
          throw new InvariantViolation(`Invalid record ${label}`);
        }

        const name = label.replace(/^:/, "");

        return {
          type: "Property",
          key: isValidJSIdentifierName(name) ? identifier(name) : literal(name),
          value: compile(value, env),
          kind: "init",
          method: false,
          shorthand: false,
          computed: false
        };
      }
    )
  };
  if (expr.extends) {
    return methodCall({ type: "Identifier", name: "Object" }, "assign", [
      { type: "ObjectExpression", properties: [] },
      compile(expr.extends, env),
      newObj
    ]);
  } else {
    return newObj;
  }
}

function compileNumber(value: number): JS.Expression {
  if (value >= 0) {
    return literal(value);
  } else {
    return {
      type: "UnaryExpression",
      operator: "-",
      prefix: true,
      argument: literal(-value)
    };
  }
}

export function compile(expr: Expression, env: Environment): JS.Expression {
  switch (expr.tag) {
    case "number":
      return compileNumber(expr.value);
    case "string":
      return literal(expr.value);
    case "vector":
      return compileVector(expr, env);
    case "record":
      return compileRecord(expr, env);
    case "identifier":
      return compileIdentifier(expr, env);
    case "conditional":
      return compileConditional(expr, env);
    case "function":
      return compileLambda(expr, env);
    case "function-call":
      return compileFunctionCall(expr, env);
    case "let-bindings":
      return compileLetBindings(expr, env);
    case "type-annotation":
      return compile(expr.value, env);
  }
}

function compileTopLevel(
  syntax: Syntax,
  env: Environment
): JS.Statement | null {
  if (isExport(syntax) || isTypeAlias(syntax)) {
    // exports are compiled at the end of the module
    return null;
  }

  let js: JS.Statement;

  if (isDefinition(syntax)) {
    js = compileDefinition(syntax, env);
  } else {
    js = {
      type: "ExpressionStatement",
      expression: compile(syntax, env)
    };
  }

  return {
    ...js,
    // Include a comment with the original source code immediately
    // before each toplevel compilation.
    leadingComments: [
      {
        type: "Block",
        value: `
${pprint(syntax, 60)}
`
      }
    ]
  };
}

function compileRuntime(env: Environment): JS.Statement | JS.ModuleDeclaration {
  return env.moduleFormat.importRuntime("env");
}

function compileExports(
  exps: SExport[],
  env: Environment
): Array<JS.Statement | JS.ModuleDeclaration> {
  const exportNames = exps.map(exp => {
    const binding = lookupBinding(exp.value.name, env);
    if (!binding || binding.source !== "module") {
      throw new Error(
        printHighlightedExpr(
          "You can only export user definitions",
          exp.value.location
        )
      );
    } else {
      return binding.jsname;
    }
  });

  if (exportNames.length > 0) {
    return [env.moduleFormat.export(exportNames)];
  } else {
    return [];
  }
}

export function moduleEnvironment(
  m: Module,
  opts: CompilerOptions = {}
): Environment {
  const moduleDefinitions = m.body
    .filter(isDefinition)
    .map(decl => decl.variable);
  const moduleBindings = moduleDefinitions.reduce(
    (d, decl) => ({
      ...d,
      [decl]: { jsname: identifierToJS(decl), source: "module" }
    }),
    {}
  );

  const primitiveBindings = mapObject(
    runtime,
    (_, key: string): EnvironmentBinding => ({
      jsname: key,
      source: "primitive"
    })
  );

  const initialEnv = {
    defs: opts.definitionContainer
      ? dynamicDefinition(opts.definitionContainer)
      : staticDefinition,
    moduleFormat: opts.esModule ? esm : cjs,
    bindings: { ...primitiveBindings, ...moduleBindings }
  };

  return initialEnv;
}

function compileModule(
  m: Module,
  includeRuntime: boolean,
  env: Environment
): JS.Program {
  return {
    type: "Program",
    sourceType: "module",
    body: [
      ...(includeRuntime ? [compileRuntime(env)] : []),
      ...maybeMap((syntax: Syntax) => compileTopLevel(syntax, env), m.body),
      ...compileExports(m.body.filter(isExport), env)
    ]
  };
}

export function compileToString(syntax: Syntax, env: Environment): string {
  const ast = compileModule({ tag: "module", body: [syntax] }, false, env);
  const code = escodegen.generate(ast, { comment: true });
  debug("jscode:", code);
  return code;
}

export function compileModuleToString(
  m: Module,
  opts: CompilerOptions = {}
): string {
  const env = moduleEnvironment(m, opts);
  const ast = compileModule(m, true, env);
  const code = escodegen.generate(ast, { comment: true });
  debug("jscode:", code);
  return code;
}
