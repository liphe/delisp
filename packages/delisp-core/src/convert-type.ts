import { capitalize } from "./utils";
import { ConvertError, parseRecord } from "./convert-utils";
import { printHighlightedExpr } from "./error-report";
import { readFromString } from "./reader";
import { last } from "./utils";

import {
  isSymbolOfName,
  ASExpr,
  ASExprList,
  ASExprMap,
  ASExprSymbol,
  ASExprVector
} from "./sexpr";

import {
  emptyRow,
  TConstant,
  TVar,
  Type,
  tApp,
  tBoolean,
  tNumber,
  tRecord,
  tMultiValuedFunction,
  tValues,
  tCases,
  tEffect,
  tString,
  tUserDefined,
  tVar,
  tVector,
  tVoid,
  tNone,
  tcArrow,
  tcStar,
  TypeSchema
} from "./types";

import { TypeWithWildcards } from "./type-wildcards";

/** Return true if a symbol is a valid name for a user defined type, false otherwise. */
export function userDefinedType(expr: ASExprSymbol): boolean {
  return /^[A-Z]/.test(expr.name);
}

/** Check if a symbol is a valid user defined type name or throw a user-friendly error otherwise. */
export function checkUserDefinedTypeName(expr: ASExprSymbol): void {
  if (!userDefinedType(expr)) {
    throw new ConvertError(
      printHighlightedExpr(
        `'${
          expr.name
        }' is not a valid name for a type. Type names should start with a capital letter. Try '${capitalize(
          expr.name
        )}'?`,
        expr.location
      )
    );
  }
}

function convertSymbol(expr: ASExprSymbol): TVar | TConstant {
  switch (expr.name) {
    case "->":
      return tcArrow;
    case "boolean":
      return tBoolean;
    case "number":
      return tNumber;
    case "string":
      return tString;
    case "void":
      return tVoid;
    case "none":
      return tNone;
    case "*":
      return tcStar;
    default:
      return userDefinedType(expr) ? tUserDefined(expr.name) : tVar(expr.name);
  }
}

function convertEffect(effects: ASExpr[]): Type {
  const labels = effects.map(e => {
    if (e.tag !== "symbol") {
      throw new ConvertError(
        printHighlightedExpr(`not a valid effect`, e.location)
      );
    }
    return e.name;
  });
  const last = labels.slice(-2);
  if (last.length === 2 && last[0] === "|") {
    return tEffect(labels.slice(0, -2), tVar(last[1]));
  } else {
    return tEffect(labels);
  }
}

function convertCases(_op: ASExpr, args: ASExpr[]): Type {
  const lastArg = last(args);
  let tail: ASExpr | undefined;

  if (lastArg && lastArg.tag === "symbol" && !lastArg.name.startsWith(":")) {
    tail = args.pop();
  }

  function parseCase(c: ASExpr): { label: string; type: Type } {
    if (c.tag === "symbol" && c.name.startsWith(":")) {
      return { label: c.name, type: tVoid };
    } else if (c.tag === "list") {
      if (c.elements.length !== 2) {
        throw new ConvertError(
          printHighlightedExpr(
            `Invalid case: expected exaclty two elements`,
            c.location
          )
        );
      }
      const [keyword, innerType] = c.elements;
      if (keyword.tag !== "symbol" || !keyword.name.startsWith(":")) {
        throw new ConvertError(
          printHighlightedExpr(
            `Invalid case: expected valid keyword`,
            keyword.location
          )
        );
      }
      return { label: keyword.name, type: convert_(innerType) };
    } else {
      throw new ConvertError(printHighlightedExpr(`Invalid case`, c.location));
    }
  }

  return tCases(args.map(parseCase), tail && convert_(tail));
}

function convertValues(expr: ASExpr): Type[] {
  // Check if X is the symbol VALUES
  if (
    expr.tag === "list" &&
    expr.elements.length > 0 &&
    isSymbolOfName(expr.elements[0], "values")
  ) {
    return expr.elements.slice(1).map(convert_);
  } else {
    return [convert_(expr)];
  }
}

function convertList(expr: ASExprList): Type {
  const [op, ...args] = expr.elements;
  if (op.tag === "symbol" && op.name === "effect") {
    return convertEffect(args);
  } else if (op.tag === "symbol" && op.name === "cases") {
    return convertCases(op, args);
  } else if (op.tag === "symbol" && op.name === "->") {
    if (args.length < 2) {
      throw new ConvertError(`Missing some arguments!`);
    }
    const fnargs = args.slice(0, -2);
    const fneffect = args[args.length - 2];
    const fnout = args[args.length - 1];
    return tMultiValuedFunction(
      fnargs.map(convert_),
      convert_(fneffect),
      tValues(convertValues(fnout))
    );
  } else {
    const opType = convert_(op);
    return tApp(opType, ...args.map(convert_));
  }
}

function convertVector(expr: ASExprVector): Type {
  if (expr.elements.length !== 1) {
    throw new ConvertError(
      printHighlightedExpr("Expected exactly 1 argument", expr.location)
    );
  }
  return tVector(convert_(expr.elements[0]));
}

function convertMap(expr: ASExprMap): Type {
  const { fields, tail } = parseRecord(expr);

  return tRecord(
    fields.map(({ label, value }) => ({
      label: label.name,
      type: convert_(value)
    })),
    tail ? convert_(tail) : emptyRow
  );
}

/* Try to convert a S-Expression into a type. */
function convert_(expr: ASExpr): Type {
  switch (expr.tag) {
    case "list":
      return convertList(expr);
    case "symbol":
      return convertSymbol(expr);
    case "vector":
      return convertVector(expr);
    case "map":
      return convertMap(expr);
    default:
      throw new ConvertError(
        printHighlightedExpr("Not a valid type", expr.location)
      );
  }
}

function convertToTypeWithWildcards(expr: ASExpr): TypeWithWildcards {
  return new TypeWithWildcards(convert_(expr));
}

export { convertToTypeWithWildcards as convert };

export function readType(source: string): TypeSchema {
  const t = convertToTypeWithWildcards(readFromString(source));
  return t.generalize();
}
