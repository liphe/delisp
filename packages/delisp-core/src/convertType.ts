import { printHighlightedExpr } from "./error-report";
import { ASExpr, ASExprList, ASExprSymbol } from "./sexpr";
import { Type } from "./types";

function convertSymbol(expr: ASExprSymbol): Type {
  switch (expr.name) {
    case "number":
      return { type: "number" };
    case "string":
      return { type: "string" };
    default:
      return { type: "type-variable", name: expr.name };
  }
}

function convertList(expr: ASExprList): Type {
  const [first, from, to] = expr.elements;

  if (first.type !== "symbol" || first.name !== "->") {
    throw new Error(printHighlightedExpr("Expected -> as first symbol", expr));
  }

  if (expr.elements.length !== 3) {
    throw new Error(printHighlightedExpr("Expected 3 elements", expr));
  }

  if (from.type !== "list") {
    throw new Error(
      printHighlightedExpr("Expected second element to be a list", expr)
    );
  }

  return {
    type: "function",
    from: from.elements.map(convert),
    to: convert(to)
  };
}

export function convert(expr: ASExpr): Type {
  switch (expr.type) {
    case "list":
      return convertList(expr);
    case "symbol":
      return convertSymbol(expr);
    default:
      throw new Error(printHighlightedExpr("Not a valid type", expr));
  }
}
