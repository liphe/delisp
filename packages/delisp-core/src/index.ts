import { convert as convertSyntax } from "./convert";

import { readFromString } from "./reader";

export { collectConvertErrors } from "./convert";

export {
  compileToString,
  compileModuleToString,
  moduleEnvironment
} from "./compiler";
export { createContext, evaluate, evaluateModule } from "./eval";
export {
  inferType,
  inferModule,
  inferExpressionInModule,
  defaultEnvironment
} from "./infer";
export { Type } from "./types";
export { printType } from "./type-printer";
export {
  isDeclaration,
  isDefinition,
  isExpression,
  isExport,
  isTypeAlias
} from "./syntax";
export { default as primitives } from "./primitives";

export { printHighlightedExpr } from "./error-report";

export function readSyntax(source: string) {
  return convertSyntax(readFromString(source));
}

export { findSyntaxByOffset, findSyntaxByRange } from "./syntax-utils";

export {
  createModule,
  readModule,
  addToModule,
  removeModuleDefinition,
  removeModuleTypeDefinition,
  moduleDefinitions,
  moduleDefinitionByName
} from "./module";

export { generateTSModuleDeclaration } from "./typescript-generation";

export { Encoder } from "./prettier";
export { pprintAs, pprint, pprintModuleAs, pprintModule } from "./printer";
