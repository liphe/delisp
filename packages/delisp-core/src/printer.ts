import {
  align,
  concat,
  Doc,
  group,
  groupalign,
  indent as indent_,
  join,
  line,
  pretty,
  space,
  text
} from "./prettier";
import { Module, Syntax } from "./syntax";

import { printType } from "./type-printer";

function indent(x: Doc, level = 2): Doc {
  return indent_(x, level);
}

function printString(str: string): Doc {
  const escaped = str.replace(/\n/g, "\\n").replace(/"/g, '\\"');
  return text(`"${escaped}"`);
}

function printIdentifier(name: string): Doc {
  return text(name);
}

function list(...docs: Doc[]): Doc {
  return concat(text("("), ...docs, text(")"));
}

function vector(...docs: Doc[]): Doc {
  return concat(text("["), ...docs, text("]"));
}

function map(...docs: Doc[]): Doc {
  return concat(text("{"), ...docs, text("}"));
}

function printBody(ss: Syntax[]): Doc {
  return concat(line, join(ss.map(print), line));
}

function print(sexpr: Syntax): Doc {
  switch (sexpr.tag) {
    case "string":
      return printString(sexpr.value);
    case "number":
      return text(String(sexpr.value));
    case "vector": {
      const args = sexpr.values.map(print);
      return group(vector(align(...args)));
    }
    case "record": {
      return group(
        map(
          align(
            join(
              sexpr.fields.map(({ label: k, value: v }) =>
                concat(text(k), space, print(v))
              ),
              line
            )
          )
        )
      );
    }
    case "identifier":
      return printIdentifier(sexpr.name);

    case "conditional":
      return group(
        list(
          concat(
            text("if"),
            space,
            align(
              print(sexpr.condition),
              print(sexpr.consequent),
              print(sexpr.alternative)
            )
          )
        ),
        // NOTE: We don't want conditionals to be too long. We use a
        // much lower value here, we'll only group them if they are
        // realluy short.
        30
      );

    case "function":
      const argNames = sexpr.lambdaList.positionalArgs.map(x => x.name);
      const singleBody = sexpr.body.length === 1;
      const doc = list(
        text("lambda"),
        space,
        group(list(align(...argNames.map(printIdentifier)))),
        indent(printBody(sexpr.body))
      );
      return singleBody ? group(doc) : doc;

    case "function-call": {
      const fn = print(sexpr.fn);
      const args = sexpr.args.map(print);
      if (args.length === 0) {
        return group(list(fn));
      } else {
        return group(list(groupalign(fn, align(...args))));
      }
    }

    case "definition":
      return group(
        list(
          text("define"),
          space,
          printIdentifier(sexpr.variable),
          indent(concat(line, print(sexpr.value)))
        )
      );

    case "export":
      return list(text("export"), space, text(sexpr.value.name));

    case "let-bindings":
      return list(
        text("let"),
        space,
        map(
          align(
            ...sexpr.bindings.map(b =>
              concat(text(b.variable.name), space, print(b.value))
            )
          )
        ),
        indent(printBody(sexpr.body))
      );

    case "type-annotation":
      return group(
        list(
          text("the"),
          space,
          text(sexpr.typeWithWildcards.print()),
          indent(concat(line, print(sexpr.value)))
        )
      );

    case "type-alias":
      return group(
        list(
          text("type"),
          space,
          text(sexpr.name),
          indent(concat(line, text(printType(sexpr.definition, false))))
        )
      );
  }
}

export function pprint(sexpr: Syntax, lineWidth: number): string {
  return pretty(print(sexpr), lineWidth);
}

export function pprintModule(m: Module, lineWidth: number): string {
  return m.body
    .map((s, i) => {
      let newlines;
      if (i > 0) {
        const end = m.body[i - 1].location.end;
        const start = s.location.start;
        const between = s.location.input.toString().slice(end, start);
        newlines = between.split("\n").length - 1;
      } else {
        newlines = 0;
      }
      const nl = newlines > 1 ? "\n" : "";
      return nl + pprint(s, lineWidth);
    })
    .join("\n");
}
