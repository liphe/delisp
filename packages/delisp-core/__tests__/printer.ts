import { readModule, readSyntax } from "../src/index";
import { pprint, pprintModule } from "../src/printer";

function pprintString(source: string, lineWidth: number = 80) {
  const syntax = readSyntax(source);
  return "\n" + pprint(syntax, lineWidth);
}

function pprintLines(source: string, lineWidth: number = 80) {
  const m = readModule(source);
  return "\n" + pprintModule(m, lineWidth);
}

function sansSpace(str: string) {
  return str && str.replace(/\s/g, "");
}

function prettySimilar(src: string) {
  expect(sansSpace(pprintLines(src))).toBe(sansSpace(src));
}

describe("Pretty Printer", () => {
  it("should insert newlines", () => {
    expect(pprintLines(`aaa bbb ccc`)).toMatchSnapshot();
  });
  it("should preserve some empty lines", () => {
    expect(
      pprintLines(`
aaa
bbb

ccc
ddd


eee
`)
    ).toMatchSnapshot();
  });

  it("should only touch spaces and newlines", () => {
    prettySimilar(`(define sqr (lambda (x) (* x x))) (export sqr)`);

    prettySimilar(
      `(lambda (name age) {:name (the string name) :age (the number age)})`
    );

    prettySimilar(
      `(define foo (lambda (a b)
         (let {ap100 (+ 100 a) bsum (fold + b 0) msg "Hello!"}
         (print msg)
         {:x ap100 :y [bsum bsum bsum]}
         ))) (export foo)`
    );

    prettySimilar(
      `(the (-> string _a {:name string :info _a}) (lambda (name info) {:name name :info info}))`
    );
  });

  it.skip("should retain partial type annotations", () => {
    prettySimilar(
      `(the (-> string number _) (lambda (name age) {:name name :age age}))`
    );
  });

  it.skip("should retain at-expressions", () => {
    prettySimilar(`@doc{This is simple annotation}`);
  });

  it("should print lambda abstractions beautifully", () => {
    expect(pprintString("(lambda (aaa bbb ccc) xxx)")).toMatchSnapshot();
    expect(
      pprintString(
        "(lambda (aaa bbb ccc ddd eee fff ggg hhh iii jjj kkk lll mmm nnn ooo ppp qqq) xxx)"
      )
    ).toMatchSnapshot();
    expect(
      pprintString(
        "(lambda (aaa bbb ccc ddd eee fff ggg hhh iii jjj kkk lll mmm nnn ooo ppp qqq rrr sss ttt uuu vvv www xx) xxx)"
      )
    ).toMatchSnapshot();
  });

  it("shold print definitions beautifully", () => {
    expect(pprintString("(define x 10)")).toMatchSnapshot();
    expect(
      pprintString(
        "(define x (lambda (aaa bbb ccc ddd eee fff ggg hhh iii jjj kkk lll mmm nnn ooo ppp qqq rrr sss ttt uuu vvv www xx) xxx))"
      )
    ).toMatchSnapshot();
  });

  it("should align nested function calls", () => {
    expect(
      pprintString(
        "(funcall aaaaa aaaaa aaaaa aaaaa aaaaa aaaaa aaaaa aaaaa (fxyz 000 111 222) aaaaa aaaaa aaaaa aaaaa aaaaa aaaaa aaaaa)"
      )
    ).toMatchSnapshot();

    expect(
      pprintString("(funcall aaaaa aaaaa aaaaa (fxyz 000 111 222))")
    ).toMatchSnapshot();

    expect(
      pprintString(
        "(funcall aaaaa aaaaa aaaaa (fxyz 000 111 222 333 444 555 666 777 888 999 aaa bbb xxx yyy zzz www uuu ttt))"
      )
    ).toMatchSnapshot();
  });

  it("should pretty print a combination of lambda and function call", () => {
    expect(
      pprintString(
        "(foo (lambda (x y z) (funcall aaaaa aaaaa aaaaa (fxyz 000 111 222 333 444 555 666 777 888 999 aaa bbb xxx yyy zzz www uuu ttt))))"
      )
    ).toMatchSnapshot();
  });

  it("should print amll real code beautifully", () => {
    expect(
      pprintString(
        `
(define bq-frob
  (lambda (x)
    (and (consp x) (or (eq (car x) *comma*) (eq (car x) *comma-atsign*)))))
`
      )
    ).toMatchSnapshot();
  });

  it("should place all arguments of a function call in the next line if necessary", () => {
    expect(
      pprintString(
        `
(this-is-a-very-long-and-ugly-function-name
 (axbxcxd (a b c d)
          (a b c d)
          (a b c d)
          (a b c d)
          (a b c d)))
`,
        40
      )
    ).toMatchSnapshot();
  });

  it("should print deeply nested function calls nicely", () => {
    expect(
      pprintString(`
(function-call-1 (function-call-2 (function-call-3 (function-call-4 (function-call-5 3
                                                                                     4
                                                                                     5
                                                                                     6
                                                                                     7))))
                 30)`)
    ).toMatchSnapshot();
  });

  it("should print let expressions nicely", () => {
    expect(pprintString(`(let {x 10 y 20} (+ x y))`)).toMatchSnapshot();
  });

  it("should print conditional expressions nicely", () => {
    expect(pprintString(`(if true 1 2)`)).toMatchSnapshot();
    expect(
      pprintString(
        `(if aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa 1 2)`
      )
    ).toMatchSnapshot();
  });

  it("should print empty vectors", () => {
    expect(pprintString(`[]`)).toMatchSnapshot();
  });

  it("should print empty records", () => {
    expect(pprintString(`{}`)).toMatchSnapshot();
  });

  it("should print records with nested expressions", () => {
    expect(
      pprintString(
        `{:xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx 20 :z (* 10 (+ 1 2))}`
      )
    ).toMatchSnapshot();
  });

  it("should some real code", () => {
    expect(
      pprintString(
        `
(define maptree
  (lambda (fn x)
    (if (atom x)
          (funcall fn x)
          (let {a (funcall fn (car x))
                d (maptree fn (cdr x))}
            (if (and (eql a (car x)) (eql d (cdr x)))
                x
                (cons a d))))))
`
      )
    ).toMatchSnapshot();
  });

  it("should print lambda with multiple forms", () => {
    expect(
      pprintString(
        `
(define hello (lambda (x)
                (print x)
                (print x)
                (print x)
                (print x)
                (print x)
                (print x)))
`
      )
    ).toMatchSnapshot();
  });

  it("should pretty print type declarations", () => {
    expect(
      pprintString(
        `(type Person {:name string :age number :books [{:name string :author string}]}) `
      )
    ).toMatchSnapshot();
  });
});
