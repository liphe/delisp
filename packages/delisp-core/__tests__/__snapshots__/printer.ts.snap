// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pretty Printer shold print definitions beautifully 1`] = `
"
(define x 10)"
`;

exports[`Pretty Printer shold print definitions beautifully 2`] = `
"
(define x
  (lambda (aaa
           bbb
           ccc
           ddd
           eee
           fff
           ggg
           hhh
           iii
           jjj
           kkk
           lll
           mmm
           nnn
           ooo
           ppp
           qqq
           rrr
           sss
           ttt
           uuu
           vvv
           www
           xx)
    xxx))"
`;

exports[`Pretty Printer should align nested function calls 1`] = `
"
(funcall aaaaa
         aaaaa
         aaaaa
         aaaaa
         aaaaa
         aaaaa
         aaaaa
         aaaaa
         (fxyz 0 111 222)
         aaaaa
         aaaaa
         aaaaa
         aaaaa
         aaaaa
         aaaaa
         aaaaa)"
`;

exports[`Pretty Printer should align nested function calls 2`] = `
"
(funcall aaaaa aaaaa aaaaa (fxyz 0 111 222))"
`;

exports[`Pretty Printer should align nested function calls 3`] = `
"
(funcall aaaaa
         aaaaa
         aaaaa
         (fxyz 0
               111
               222
               333
               444
               555
               666
               777
               888
               999
               aaa
               bbb
               xxx
               yyy
               zzz
               www
               uuu
               ttt))"
`;

exports[`Pretty Printer should insert newlines 1`] = `
"
aaa
bbb
ccc"
`;

exports[`Pretty Printer should place all arguments of a function call in the next line if necessary 1`] = `
"
(this-is-a-very-long-and-ugly-function-name
 (axbxcxd (a b c d)
          (a b c d)
          (a b c d)
          (a b c d)
          (a b c d)))"
`;

exports[`Pretty Printer should preserve some empty lines 1`] = `
"
aaa
bbb

ccc
ddd

eee"
`;

exports[`Pretty Printer should pretty print a combination of lambda and function call 1`] = `
"
(foo (lambda (x y z)
       (funcall aaaaa
                aaaaa
                aaaaa
                (fxyz 0
                      111
                      222
                      333
                      444
                      555
                      666
                      777
                      888
                      999
                      aaa
                      bbb
                      xxx
                      yyy
                      zzz
                      www
                      uuu
                      ttt))))"
`;

exports[`Pretty Printer should print amll real code beautifully 1`] = `
"
(define bq-frob
  (lambda (x)
    (and (consp x) (or (eq (car x) *comma*) (eq (car x) *comma-atsign*)))))"
`;

exports[`Pretty Printer should print conditional expressions nicely 1`] = `
"
(if true 1 2)"
`;

exports[`Pretty Printer should print conditional expressions nicely 2`] = `
"
(if aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
    1
    2)"
`;

exports[`Pretty Printer should print deeply nested function calls nicely 1`] = `
"
(function-call-1 (function-call-2 (function-call-3 (function-call-4
                                                    (function-call-5 3 4 5 6 7))))
                 30)"
`;

exports[`Pretty Printer should print lambda abstractions beautifully 1`] = `
"
(lambda (aaa bbb ccc) xxx)"
`;

exports[`Pretty Printer should print lambda abstractions beautifully 2`] = `
"
(lambda (aaa bbb ccc ddd eee fff ggg hhh iii jjj kkk lll mmm nnn ooo ppp qqq)
  xxx)"
`;

exports[`Pretty Printer should print lambda abstractions beautifully 3`] = `
"
(lambda (aaa
         bbb
         ccc
         ddd
         eee
         fff
         ggg
         hhh
         iii
         jjj
         kkk
         lll
         mmm
         nnn
         ooo
         ppp
         qqq
         rrr
         sss
         ttt
         uuu
         vvv
         www
         xx)
  xxx)"
`;

exports[`Pretty Printer should print lambda with multiple forms 1`] = `
"
(define hello
  (lambda (x)
    (print x)
    (print x)
    (print x)
    (print x)
    (print x)
    (print x)))"
`;

exports[`Pretty Printer should print let expressions nicely 1`] = `
"
(let ((x 10)
      (y 20))
  (+ x y))"
`;

exports[`Pretty Printer should print records with nested expressions 1`] = `
"
{xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx 20
 z (* 10 (+ 1 2))}"
`;

exports[`Pretty Printer should some real code 1`] = `
"
(define maptree
  (lambda (fn x)
    (if (atom x)
        (funcall fn x)
        (let ((a (funcall fn (car x)))
              (d (maptree fn (cdr x))))
          (if (and (eql a (car x)) (eql d (cdr x)))
              x
              (cons a d))))))"
`;
